src/actions/copy_object.rs

```diff
@@ -0,0 +1,151 @@
+use std::time::Duration;
+
+use time::OffsetDateTime;
+use url::Url;
+
+use super::S3Action;
+use crate::actions::Method;
+use crate::signing::sign;
+use crate::{Bucket, Credentials, Map};
+
+/// Create a copy of an object that is already stored in S3, using a `PUT` request.
+///
+/// Find out more about CopyObject from the [AWS API Reference][api]
+///
+/// [api]: https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html
+#[derive(Debug, Clone)]
+pub struct CopyObject<'a> {
+    bucket: &'a Bucket,
+    credentials: Option<&'a Credentials>,
+    src_object: &'a str,
+    dst_object: &'a str,
+
+    query: Map<'a>,
+    headers: Map<'a>,
+}
+
+impl<'a> CopyObject<'a> {
+    #[inline]
+    pub fn new(
+        bucket: &'a Bucket,
+        credentials: Option<&'a Credentials>,
+        src_object: &'a str,
+        dst_object: &'a str,
+    ) -> Self {
+        let mut query = Map::new();
+        query.insert(
+            "x-amz-copy-source",
+            format!("{}/{}", bucket.name(), src_object),
+        );
```

Comment:
Should this use `SortingIterator` while signing to save a Map allocation, like [here](https://github.com/paolobarbolini/rusty-s3/blob/main/src/actions/delete_objects.rs#L133)?

Reply:
By prefixing the current bucket name, this action does not allow copying from a different bucket.

Reply:
> Should this use `SortingIterator` while signing to save a Map allocation, like [here](https://github.com/paolobarbolini/rusty-s3/blob/main/src/actions/delete_objects.rs#L133)?

Sure, sounds like a good use for it.

> By prefixing the current bucket name, this action does not allow copying from a different bucket.

We could take both a `src_bucket` and a `dest_bucket`

Reply:
Taking a `Bucket` is unnecessarily involved, as you need to pass connection details for bucket construction which will go completely unused. How about taking the raw source string, and having two separate functions on `Bucket`, e.g.:

- `Bucket::copy_object` would prepend the current bucket name,
- `Bucket::copy_object_from_bucket` would pass the source string as-is, and require the caller to prepend a (potentially different) bucket name

Reply:
Sounds good